bind-2/bind/doc/man/getnameinfo.3
.\" $Id: getnameinfo.3,v 1.1.1.1 1999/10/04 22:26:17 wsanchez Exp $
.\"
.\"Copyright (c) 1998,1999 by Internet Software Consortium
.\"
.\"Permission to use, copy, modify, and distribute this software for any
.\"purpose with or without fee is hereby granted, provided that the above
.\"copyright notice and this permission notice appear in all copies.
.\"
.\"THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM DISCLAIMS
.\"ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
.\"OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL INTERNET SOFTWARE
.\"CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
.\"DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
.\"PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
.\"ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
.\"SOFTWARE.
.\"
.Dd January 11, 1999
.Dt GETRNAMEINFO @LIB_NETWORK_EXT@
.Sh NAME
.Nm getnameinfo
.Nd address-to-name translation in protocol-independent manner
.Sh SYNOPSIS
.Fd #include <sys/socket.h>
.Fd #include <netdb.h>
.Ft int
.Fn getnameinfo "const struct sockaddr *sa" "socklen_t salen" \
"char *host" "size_t hostlen" "char *serv" "size_t servlen" "int flags"
.Sh DESCRIPTION
The
.Fn getnameinfo
function is defined for protocol-independent address-to-nodename translation.
It performs functionality of 
.Xr gethostbyaddr @LIB_NETWORK_EXT@
and
.Xr getservbyport @LIB_NETWORK_EXT@
in more sophisticated manner.
.Pp
The
.Fa sa
arguement is a pointer to a generic socket address structure of size
.Fa salen .
The arguements
.Fa host
and
.Fa serv
are pointers to buffers to hold the return values.
Their sizes are specified by
.Fa hostlen
and
.Fa servlen
repectively.
Either
.Fa host
or
.Fa serv
may be
.Dv NULL
if the hostname  or service name is not required.
.Pp
The 
.Fa flags
arguement modifies the behaviour of 
.Fn getnameinfo
as follows:
.Pp
If 
.Dv NI_NOFQDN
is set only the unqualified hostname is returned for local fully
qualified names.
.Pp
If
.Dv NI_NUMERICHOST
is set then the numeric form of the hostname is returned.
.Pp
If
.Dv NI_NAMEREQD
is set, then a error is returned if the hostname cannot be looked up.
.Pp
If
.Dv NI_NUMERICSERV
is set then the service is returned in numeric form.
.Pp
If
.Dv NI_DGRAM
is set then the service is UDP based rather than TCP based.
.Sh SEE ALSO
.Xr getaddrinfo @LIB_NETWORK_EXT@ ,
.Xr gethostbyaddr @LIB_NETWORK_EXT@ ,
.Xr getservbyport @LIB_NETWORK_EXT@ ,
.Xr hosts @FORMAT_EXT@ ,
.Xr services @FORMAT_EXT@ ,
.Xr hostname @DESC_EXT@ ,
.Xr named @SYS_OPS_EXT@
.Pp
R. Gilligan, S.  Thomson, J. Bound, and W. Stevens,
``Basic Socket Interface Extensions for IPv6,'' RFC2133, April 1997.
.Sh STANDARDS
The
.Fn getaddrinfo
function is defined IEEE POSIX 1003.1g draft specification,
and documented in ``Basic Socket Interface Extensions for IPv6''
.Pq RFC2133 .


bind-2/bind/src/lib/irs/getnameinfo.c
/*
 * Issues to be discussed:
 * - Thread safe-ness must be checked
 * - Return values.  There seems to be no standard for return value (RFC2133)
 *   but INRIA implementation returns EAI_xxx defined for getaddrinfo().
 */

/*
 * Copyright (C) 1995, 1996, 1997, and 1998 WIDE Project.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *    This product includes software developed by WIDE Project and
 *    its contributors.
 * 4. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <port_before.h>

#include <sys/types.h>
#include <sys/socket.h>

#include <netinet/in.h>
#include <arpa/nameser.h>
#include <arpa/inet.h>

#include <netdb.h>
#include <resolv.h>
#include <string.h>

#include <port_after.h>

#define SUCCESS 0
#define ANY 0
#define YES 1
#define NO  0

/*
 * Note that a_off will be dynamically adjusted so that to be consistent
 * with the definition of sockaddr_in{,6}.
 * The value presented below is just a guess.
 */
static struct afd {
	int a_af;
	int a_addrlen;
	int a_socklen;
	int a_off;
} afdl [] = {
	/* first entry is linked last... */
	{PF_INET, sizeof(struct in_addr), sizeof(struct sockaddr_in),
		4 /*XXX*/},
	{PF_INET6, sizeof(struct in6_addr), sizeof(struct sockaddr_in6),
		8 /*XXX*/},
	{0, 0, 0},
};

struct sockinet {
	u_char	si_len;
	u_char	si_family;
	u_short	si_port;
};

#define ENI_NOSOCKET 	0
#define ENI_NOSERVNAME	1
#define ENI_NOHOSTNAME	2
#define ENI_MEMORY	3
#define ENI_SYSTEM	4
#define ENI_FAMILY	5
#define ENI_SALEN	6

int
getnameinfo(sa, salen, host, hostlen, serv, servlen, flags)
	const struct sockaddr *sa;
	size_t salen;
	char *host;
	size_t hostlen;
	char *serv;
	size_t servlen;
	int flags;
{
	struct afd *afd;
	struct servent *sp;
	struct hostent *hp;
	u_short port;
	int family, len, i;
	char *addr, *p;
	u_long v4a;
	u_char pfx;
	static int firsttime = 1;
	static char numserv[512];
	static char numaddr[512];


	/* dynamically adjust a_off */
	if (firsttime) {
		struct afd *p;
		u_char *q;
		struct sockaddr_in sin;
		struct sockaddr_in6 sin6;

		for (p = &afdl[0]; p->a_af; p++) {
			switch (p->a_af) {
			case PF_INET:
				q = (u_char *)&sin.sin_addr.s_addr;
				p->a_off = q - (u_char *)&sin;
				break;
			case PF_INET6:
				q = (u_char *)&sin6.sin6_addr.s6_addr;
				p->a_off = q - (u_char *)&sin6;
				break;
			default:
				break;
			}
		}
		firsttime = 0;
	}

	if (sa == NULL)
		return ENI_NOSOCKET;

#ifdef HAVE_SA_LEN
	len = sa->sa_len;
	if (len != salen) return ENI_SALEN;
#endif
	
	family = sa->sa_family;
	for (i = 0; afdl[i].a_af; i++)
		if (afdl[i].a_af == family) {
			afd = &afdl[i];
			goto found;
		}
	return ENI_FAMILY;
	
 found:
	if (salen != afd->a_socklen) return ENI_SALEN;
	
	port = ((struct sockinet *)sa)->si_port; /* network byte order */
	addr = (char *)sa + afd->a_off;

	if (serv == NULL || servlen == 0) {
		/* what we should do? */
	} else if (flags & NI_NUMERICSERV) {
		snprintf(numserv, strlen(numserv), "%d", ntohs(port));
		if (strlen(numserv) > servlen)
			return ENI_MEMORY;
		strcpy(serv, numserv);
	} else {
		sp = getservbyport(port, (flags & NI_DGRAM) ? "udp" : "tcp");
		if (sp) {
			if (strlen(sp->s_name) > servlen)
				return ENI_MEMORY;
			strcpy(serv, sp->s_name);
		} else
			return ENI_NOSERVNAME;
	}

	switch (sa->sa_family) {
	case AF_INET:
		v4a = ((struct sockaddr_in *)sa)->sin_addr.s_addr;
		if (IN_MULTICAST(v4a) || IN_EXPERIMENTAL(v4a))
			flags |= NI_NUMERICHOST;
		v4a >>= IN_CLASSA_NSHIFT;
		if (v4a == 0 || v4a == IN_LOOPBACKNET)
			flags |= NI_NUMERICHOST;			
		break;
	case AF_INET6:
		pfx = ((struct sockaddr_in6 *)sa)->sin6_addr.s6_addr[0];
		if (pfx == 0 || pfx == 0xfe || pfx == 0xff)
			flags |= NI_NUMERICHOST;
		break;
	}
	if (host == NULL || hostlen == 0) {
		/* what should we do? */
	} else if (flags & NI_NUMERICHOST) {
		if (inet_ntop(afd->a_af, addr, numaddr, sizeof(numaddr))
		    == NULL)
			return ENI_SYSTEM;
		if (strlen(numaddr) > hostlen)
			return ENI_MEMORY;
		strcpy(host, numaddr);
	} else {
		hp = gethostbyaddr(addr, afd->a_addrlen, afd->a_af);

		if (hp) {
			if (flags & NI_NOFQDN) {
				p = strchr(hp->h_name, '.');
				if (p) *p = '\0';
			}
			if (strlen(hp->h_name) > hostlen)
				return ENI_MEMORY;
			strcpy(host, hp->h_name);
		} else {
			if (flags & NI_NAMEREQD)
				return ENI_NOHOSTNAME;
			if (inet_ntop(afd->a_af, addr, numaddr, sizeof(numaddr))
			    == NULL)
				return ENI_NOHOSTNAME;
			if (strlen(numaddr) > hostlen)
				return ENI_MEMORY;
			strcpy(host, numaddr);
		}
	}
	return SUCCESS;
}
